### PR TITLE
feat: implement bound decls in tasks and workflows in experimental parser.

### PR DESCRIPTION
**Pull request by peterhuene**
This commit implements bound declarations in tasks and workflows for the experimental parser.

As bound declarations support expressions, this also implements support for expressions, including in string interpolations.

A basic Pratt parser is implemented in the `expr_with_precedence` function.

Before submitting this PR, please make sure:

- [x] You have added a few sentences describing the PR here.
- [x] You have added yourself or the appropriate individual as the assignee.
- [x] You have added at least one relevant code reviewer to the PR.
- [x] Your code builds clean without any errors or warnings.
- [x] You have added tests (when appropriate).
- [x] You have updated the README or other documentation to account for these
      changes (when appropriate).
- [x] You have added an entry to the relevant `CHANGELOG.md` (see
      ["keep a changelog"] for more information).
- [x] Your commit messages follow the [conventional commit] style.
- [x] Your changes are squashed into a single commit (unless there is a _really_
      good, articulated reason as to why there should be more than one).

[conventional commit]: https://www.conventionalcommits.org/en/v1.0.0/#summary
["keep a changelog"]: https://keepachangelog.com/en/1.0.0/

### CODE REVIEW

File: wdl-grammar/src/experimental/grammar/v1.rs [hunk 1]
```diff
@@ -890,3 +1144,419 @@ fn call_statement(parser: &mut Parser<'_>, _marker: Marker) -> Result<(), (Marke
...
+        }
+    }
+
+    Ok(lhs)
+}
+
+/// Parses an atomic expression such as a literal.
+///
+/// Due to the WDL grammar having an ambiguity between parenthesized expressions
+/// and pair literals, this function also handles the former.
```
Comment (a-frantz) [inline wdl-grammar/src/experimental/grammar/v1.rs:1279]:
```suggestion
/// Due to the WDL grammar having an ambiguity between parenthesized expressions
/// and pair literals, this function handles the former in addition to pair literals.
```
I initially misread this comment. It was grammatically correct, but I found it confusing. I think this is a bit better.

File: wdl-grammar/tests/parsing/name-ref/source.wdl [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# This is a test of a name references.
+
+version 1.1
+
+task test {
+    Integer x = 0
+    Integer y = -x
+    Integer z = x + y * x
```
Comment (a-frantz) [inline wdl-grammar/tests/parsing/name-ref/source.wdl:8]:
```suggestion
    Int x = 0
    Int y = -x
    Int z = x + y * x
```
I think this is what was meant. What is currently being parsed is a custom type named `Integer` (that doesn't exist).

File: wdl-grammar/tests/parsing/postfix-exprs/source.wdl [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# This is a test of postfix expressions
+
+version 1.1
+
+task test {
+    Integer a = min(0, 1)
+    Integer b = min(max(100, a), 10)
+    Array[String] c = ["a", "b", "c"]
+    String d = d[a + b]
```
Comment (a-frantz) [inline wdl-grammar/tests/parsing/postfix-exprs/source.wdl:9]: `d` is undefined when used on the RHS.

File: wdl-grammar/tests/parsing/postfix-exprs/source.wdl [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# This is a test of postfix expressions
+
+version 1.1
+
+task test {
+    Integer a = min(0, 1)
+    Integer b = min(max(100, a), 10)
```
Comment (a-frantz) [inline wdl-grammar/tests/parsing/postfix-exprs/source.wdl:7]:
```suggestion
    Int a = min(0, 1)
    Int b = min(max(100, a), 10)
```

File: wdl-grammar/tests/parsing/precedence/source.wdl [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# This is a test of operator precedence
+
+version 1.1
+
+task test {
+    Boolean a = 0 || 1 && 1 == 0 != 1 < 0 <= 1 > 0 >= 1 + 2 - 3 * 4 / 5 % 6
+    Integer b = (1 + 2) - (3 * 4) / (5 % 6)
```
Comment (a-frantz) [inline wdl-grammar/tests/parsing/precedence/source.wdl:7]:
```suggestion
    Boolean a = 0 || 1 && 1 == 0 != 1 < 0 <= 1 > 0 >= 1 + 2 - 3 * 4 / 5 % 6
    Int b = (1 + 2) - (3 * 4) / (5 % 6)
```

File: wdl-grammar/tests/parsing/precedence/source.wdl [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# This is a test of operator precedence
```
Comment (a-frantz) [inline wdl-grammar/tests/parsing/precedence/source.wdl:1]: So in the Rust code I saw assignment and calculation of precedence. But am I correct that it isn't tested or reflected at this stage of the parse implementation? `source.tree` is just in the order it was found in the file, and doesn't have precedence information.

File: wdl-grammar/tests/parsing/prefix-exprs/source.wdl [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# This is a test of prefix expressions.
+
+version 1.1
+
+task test {
+    Boolean a = !true
+    Boolean b = !!!!!!!!!false
+    Int c = -2
+    Float d = -2.0
+    Int e = ---------42
```
Comment (a-frantz) [inline wdl-grammar/tests/parsing/prefix-exprs/source.wdl:10]: Is this allowed? (I haven't checked the spec) I would hope/expect an error will be thrown for this at some stage.

File: wdl-grammar/tests/parsing/name-ref/source.wdl (hunk 2 above)
Comment (peterhuene) [reply wdl-grammar/tests/parsing/name-ref/source.wdl:8]: Doh! `Integer` is legal an would be interpreted as a custom type name here; this is going to be a recurring problem in my WDL journey, I think 😭

File: wdl-grammar/tests/parsing/postfix-exprs/source.wdl (hunk 4 above)
Comment (peterhuene) [reply wdl-grammar/tests/parsing/postfix-exprs/source.wdl:7]: 😭

File: wdl-grammar/tests/parsing/precedence/source.wdl (hunk 5 above)
Comment (peterhuene) [reply wdl-grammar/tests/parsing/precedence/source.wdl:7]: I'm sensing a pattern...

File: wdl-grammar/tests/parsing/precedence/source.wdl (hunk 6 above)
Comment (peterhuene) [reply wdl-grammar/tests/parsing/precedence/source.wdl:1]: The precedence is captured in the depth of the tree, though. It just happens that the expressions are in lowest-to-highest order, so the tree shape matches.

I'll add a reverse test.

File: wdl-grammar/tests/parsing/prefix-exprs/source.wdl (hunk 7 above)
Comment (peterhuene) [reply wdl-grammar/tests/parsing/prefix-exprs/source.wdl:10]: It should be, it's just a series of negation expressions that are right-associative, no different from an explicit parenthesized expression of `-(-42)`, same [as in Rust](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6b6ed2a2ac3075e63933fc0ee7366cd0).

Comment (a-frantz) [reply wdl-grammar/tests/parsing/prefix-exprs/source.wdl:10]: Huh. Learned something new. IDK why, but I was working under the assumption that `-42` is the same as `--42` is the same as `---42` and so on. But in hindsight that doesn't make much sense. Ignore this then!

**Pull request merged**